test/integration/targets/ansible-galaxy-collection/library/setup_collections.py

```diff
@@ -241,7 +245,14 @@ def run_module():
 
     pool = threading.Pool(4)
     publish_func = partial(publish_collection, module)
-    result['results'] = pool.map(publish_func, module.params['collections'])
+    try:
+        result['results'] = pool.map_async(
+            publish_func, module.params['collections'],
+        ).get(timeout=COLLECTIONS_BUILD_AND_PUBLISH_TIMEOUT)
+    except TimeoutError as timeout_err:
+        module.fail_json(
+            'Timed out waiting for collections to be provisioned.',
+        )
```

Comment:
Instead of failing outright, should we perhaps retry one time first?

Reply:
> Instead of failing outright, should we perhaps retry one time first?

That was my first though, actually, but then it'd require rewriting `publish_collection`. This is because the current implementation uses a hardcoded temporary directory name and Python threads are unkillable. The timeout simply means "we've got tired of waiting and want to get back to the control flow here" but the threads spawned earlier keep being run and are still stuck somewhere in the `ansible_module.run_command()` invocation. If we were to implement retries, we'd need to make sure to create these dirs with unique names on each invocation or decouple building a collection artifact from publishing it and retry publishing specifically.
Either way, that "frozen" `ansible-galaxy` invocation is not deterministically killable because `run_command()` doesn't have an ability to set timeout, nor does it expose granular subprocess control — we can't even get a PID to kill it.

Though, if we agree that this is the way to go, I'll go ahead and explore it.

I also wonder if there's a way to inspect whether Pulp has been fully initialized to ensure that by the time this module is run, that instance is actually accessible/usable.

Reply:
I think it's worth looking into, but in the meantime this will already be an improvement.

Reply:
Okay, I'll merge this and try improving it in a follow-up.